pkg/controller/cloneset/sync/cloneset_scale.go

```diff
@@ -403,3 +421,164 @@ func (r *realControl) choosePodsToDelete(cs *appsv1alpha1.CloneSet, totalDiff in
 
 	return podsToDelete
 }
+
+func getInstanceIDsFromPods(pods []*v1.Pod) sets.String {
+	ins := sets.NewString()
+	for _, pod := range pods {
+		if id := clonesetutils.GetInstanceID(pod); id != "" {
+			ins.Insert(id)
+		}
+	}
+	return ins
+}
+
+func getUselessPVCs(pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim) map[string]*v1.PersistentVolumeClaim {
+	activeIDs := getInstanceIDsFromPods(pods)
+
+	useless := map[string]*v1.PersistentVolumeClaim{}
+	for _, pvc := range pvcs {
+		id := clonesetutils.GetInstanceID(pvc)
+		if id != "" && !activeIDs.Has(id) {
+			useless[id] = pvc
+		}
+	}
+	return useless
+}
+
+func (r *realControl) cleanupPVCs(cs *appsv1alpha1.CloneSet, uselessPVCs map[string]*v1.PersistentVolumeClaim) (bool, error) {
+	var modified bool
+
+	inactivePods, err := getInactivePods(r.Client, cs)
+	if err != nil {
+		klog.Errorf("Could not get cloneset %s owned inactive pods", clonesetutils.GetControllerKey(cs))
+		return modified, err
+	}
+
+	// There are two scenarios to clean up the pvc:
+	// 1. If pvc belongs to pod in terminating state, set ownerReference of the
+	//    pod to the pvc instead of the cloneset.
+	// 2. if pvc belongs to already cleaned up pods or in Succeeded/Failed state,
+	//    just delete the pvc.
+
+	for _, pod := range inactivePods {
+		instanceID := clonesetutils.GetInstanceID(pod)
+		if pvc, ok := uselessPVCs[instanceID]; ok && clonesetutils.IsTerminating(pod) {
+			if updateClaimOwnerRefToPod(pvc, cs, pod) {
+				if modified, err := r.updatePVC(cs, pvc); err != nil && !errors.IsNotFound(err) {
+					return modified, err
+				}
+				modified = true
+			}
+			// Left pvcs will be deleted directly.
+			delete(uselessPVCs, instanceID)
+		}
+	}
+
+	for _, pvc := range uselessPVCs {
+		if clonesetutils.IsTerminating(pvc) {
```

Comment:
I don’t think it’s necessary to getInactivePods in the front, just get the pod here, update the owner ref if it exists, delete it if it doesn’t exist.